src/datasets/arrow_dataset.py

```diff
@@ -589,68 +503,69 @@ def save_to_disk(self, dataset_path: str, fs=None):
         else:
             fs = fsspec.filesystem("file")
+            cache_files_paths = [Path(cache_filename) for cache_filename in self.cache_files]
+            # Check that the dataset doesn't overwrite iself. It can cause a permission error on Windows and a segfault on linux.
+            if Path(dataset_path).joinpath(DATASET_ARROW_FILENAME) in cache_files_paths:
+                raise PermissionError(
+                    f"Tried to overwrite {Path(dataset_path).joinpath(DATASET_ARROW_FILENAME)} but a dataset can't overwrite itself."
+                )
+            if Path(dataset_path).joinpath(DATASET_INDICES_FILENAME) in cache_files_paths:
+                raise PermissionError(
+                    f"Tried to overwrite {Path(dataset_path).joinpath(DATASET_INDICES_FILENAME)} but a dataset can't overwrite itself."
+                )
 
-        # create temporary directory for saving
-        with tempfile.TemporaryDirectory() as temp_dataset_path:
-            fs.makedirs(dataset_path, exist_ok=True)
-
-            # Write indices if needed
-            if self._indices is not None:
-                if not self._indices_data_files:
-                    cache_file_name = os.path.join(temp_dataset_path, "indices.arrow")
-                    with ArrowWriter(path=cache_file_name) as writer:
-                        writer.write_table(self._indices)
-                        writer.finalize()
-                    self._indices_data_files = [{"filename": cache_file_name}]
-            # Write dataset if needed
-            if not self._data_files or any(len(h["transforms"]) > 0 for h in self._inplace_history):
-                cache_file_name = os.path.join(temp_dataset_path, "dataset.arrow")
-                with ArrowWriter(path=cache_file_name) as writer:
-                    writer.write_table(self._data)
+        # Get json serializable state
+        state = {
+            key: self.__dict__[key]
+            for key in [
+                "_fingerprint",
+                "_format_columns",
+                "_format_kwargs",
+                "_format_type",
+                "_indexes",
+                "_output_all_columns",
+                "_split",
+            ]
+        }
+        state["_data_files"] = [{"filename": DATASET_ARROW_FILENAME}]
+        state["_indices_files"] = [{"filename": DATASET_INDICES_FILENAME}] if self._indices is not None else None
+        for k in state["_format_kwargs"].keys():
+            try:
+                json.dumps(state["_format_kwargs"][k])
+            except TypeError as e:
+                raise TypeError(str(e) + f"\nThe format kwargs must be JSON serializable, but key '{k}' isn't.")
+
+        # Get json serializable dataset info
+        dataset_info = asdict(self._info)
+
+        # Save dataset + indices + state + info
+        fs.makedirs(dataset_path, exist_ok=True)
+        with fs.open(Path(dataset_path).joinpath(DATASET_ARROW_FILENAME).as_posix(), "wb") as dataset_file:
```

Comment:
optional nit, pathlib overloads `/`, so that you can do `(mydirectory / myfile).as_posix()`, it just expects the left element to be a `pathlib.Path`.

Reply:
tbh you can also just `Path(mydir, myfile)` which is shorter and as explicit. not as fun as using the `/` to visually signify fs hierarchy, but has its merits 😛